spec/index.bs

```diff
@@ -1894,6 +1900,111 @@ dictionary GPUVertexAttributeDescriptor {
 };
 </script>
 
+## {{GPUPipelineBase/getBindGroupLayout()|getBindGroupLayout}} ## {#get-bind-group-layout}
+
+<dl dfn-type="method" dfn-for="GPUPipelineBase">
+    : <dfn>getBindGroupLayout(unsigned long index)</dfn>
+    ::
+        <div algorithm="GPUProgrammablePipeline.getBindGroupLayout(unsigned long index)">
+            1. If |index| is greater or equal to {{GPULimits/maxBindGroups}} :
+
+                1. Throw a {{RangeError}}.
+
+            1. If |this|.getBindGroupLayout has already been called with |index|:
+
+                1. Return the same object.
+
+            1. If |this| is an error:
+
+                1. Return an error {{GPUBindGroupLayout}}.
+
+            1. Let |bindings| be an empty sequence.
+            1. For each {{GPUProgrammableStageDescriptor}} |stage| in the descriptor for |this|:
+
+                1. For each resource |resource| in |stage|.{{GPUProgrammableStageDescriptor/module}}
+                    for the |stage|.{{GPUProgrammableStageDescriptor/entryPoint}} entry-point:
+
+                    1. If |resource|'s "group" decoration is not equal to |index|:
+
+                        1. Continue to the next loop iteration over resources.
+
+                    1. Let |binding| be a new {{GPUBindGroupLayoutBinding}}.
+                    1. Set |binding|.{{GPUBindGroupLayoutBinding/binding}} to |resource|'s "binding"
+                        decoration.
+                    1. Set |binding|.{{GPUBindGroupLayoutBinding/visibility}} to
```

Comment:
why is it visible to all stages? if we are introspecting the shader, we should know the stages that this binding should be visible in

Reply:
This helps make the default BGLs from different pipelines be more likely to be compatible even if slightly less optimal.

Reply:
oh god, that's a slippery slope. You are trading one aspect of performance (cost to switch pipeline layouts) to another (cost of binding a descriptor to more stages that it needs to). We shouldn't be making this trade-off implicitly, and we shouldn't even *try* to make them more compatible. The use case here is (btw, please do explicitly specify what the use case is, somewhere), as far as I see, having:
  - beginner user not wanting to care about pipeline compatibility
  - conscious user having a unique pipeline that doesn't need to be compatible with anything

In both cases, if they need to care about pipeline switches, they need to use pipeline layouts explicitly.

Reply:
I'm also inclined not to include all visibilities.

I don't think it would have much of an impact on usage (I think you would rarely explicitly depend on the layouts matching), so it's really just for pipeline switching performance, right?